extension-workspace-references.md

```diff
@@ -0,0 +1,115 @@
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate server-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace references exporting support.
+   */
+  xworkspaceReferencesProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace references request is sent from the client to the server to export project-wide references to dependencies. That is, the response strictly returns references in the project to symbols defined in dependencies.
+
+_Request_
+* method: 'workspace/xreferences'
+* params: `WorkspaceReferencesParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace References Request.
+ */
+interface WorkspaceReferencesParams {
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace where the `symbol` has been referenced.
+	 */
+	reference: Location;
+
+	/**
+	 * Metadata information describing the symbol being referenced.
+	 */
+	symbol: ReferenceSymbolInformation;
+}
+```
+* error: code and message set in case an exception happens during the workspace references request.
+
+Where `ReferenceSymbolInformation` is defined as follows:
+
+```typescript
+/**
+ * Represents information about a programming construct like a variable, class,
+ * interface etc that has a reference to it. Effectively, it contains data similar
+ * to SymbolInformation except all fields are optional and a metadata field is
+ * present for language-specific data.
+ *
+ * ReferenceSymbolInformation does NOT always uniquely identify a symbol. The
+ * caller should effectively consider the returned information to be
+ * information about the symbol which _generally_ (but NOT always) identifies a
+ * single symbol.
+ */
+interface ReferenceSymbolInformation {
+    /**
+     * The name of this symbol (same as `SymbolInformation.name`).
+     */
+    name?: string;
+
+    /**
+     * The kind of this symbol (same as `SymbolInformation.kind`).
+     */
+    kind?: number;
+
+    /**
+     * The URI of this symbol (same as `SymbolInformation.location.uri`).
+     */
+    uri?: string;
+
+    /**
+     * The name of the symbol containing this symbol (same as `SymbolInformation.containerName`).
+     */
+    containerName?: string;
+
+    /**
+     * The repository URI that the symbol is defined in, e.g. `github.com/golang/go`
+     * or `bitbucket.org/jespern/django-piston`.
+     */
+    repo?: string;
```

Comment:
Okay, so I chatted with @akhleung in person about the Java side of things here since I don't 100% understand what sort of metadata Java has present about a given dependency. That led me to find quite a lot of useful information (see http://stackoverflow.com/a/3724851/5905344) My current understanding is that:

1. Maven/Gradle/Ant would all have some type of globally unique identifier (e.g. maven groupId + artifactId + version) but in most cases probably not an easy, or sometimes even not possible, way to map back to a repository.
2. The actual fields that each build tool would use to uniquely identify a repo, it sounds like, would vary quite a bit in terms of naming schemes.